vm/natives.c

```diff
@@ -1138,5 +1138,393 @@ const AsebaNativeFunctionDescription AsebaNativeDescription_rand =
+{
+	"deque.get",
+	"fill dest from src at index",
+	{
+		{ -1, "dest" },
+		{ -2, "src" },
+		{  1, "index" },
+		{ 0, 0 }
+	}
+};
+
+void AsebaNative_deqset(AsebaVMState *vm)
+{
+	// variable pos
+	uint16 dest = AsebaNativePopArg(vm);
+	uint16 src = AsebaNativePopArg(vm);
+	uint16 index = AsebaNativePopArg(vm);
+	uint16 index_val = vm->variables[index];
+	
+	// variable size
+	uint16 dest_length = AsebaNativePopArg(vm);
+	uint16 src_length = AsebaNativePopArg(vm);
+
+	// Infer deque parameters
+	uint16 dq_start = vm->variables[dest + 1];
+	uint16 dq_capacity = dest_length - 2;
+	
+	// Check for exception
+	if (vm->variables[index] < 0 || vm->variables[index] > dest_length - 2 - 1
+		|| dest_length > vm->variables[src] - index_val)
+	{
+		vm->flags = ASEBA_VM_STEP_BY_STEP_MASK;
+		AsebaSendMessage(vm, ASEBA_MESSAGE_ARRAY_ACCESS_OUT_OF_BOUNDS, &(vm->pc), sizeof(vm->pc));
+		return;
+	}
+	
+	// Copy elements into deque
+	uint16 i;
+	
+	for (i = 0; i < src_length; i++)
+	{
+		vm->variables[dest + 2 + ((dq_start + index_val + i) % dq_capacity)] = vm->variables[src++];
+	}
+}
+
+const AsebaNativeFunctionDescription AsebaNativeDescription_deqset =
+{
+	"deque.set",
+	"copies src to dest at index",
+	{
+		{ -1, "dest" },
+		{ -2, "src" },
+		{  1, "index" },
+		{ 0, 0 }
+	}
+};
+	
+static void _AsebaNative_deqinsert(AsebaVMState *vm, uint16 dest, uint16 src, uint16 index_val, uint16 dest_length, uint16 src_length)
+{
+	// infer deque parameters
+	uint16 dq_size = vm->variables[dest];
+	uint16 dq_start = vm->variables[dest + 1];
+	uint16 dq_capacity = dest_length - 2;
+	
+	// Insert src elements as a block
+	// if in left half, shift prefix elements left
+	if (index_val < dq_size / 2)
+	{
+		vm->variables[dest + 1] = dq_start = (dq_start == 0) ? dq_capacity - src_length : (dq_start - src_length + dq_capacity) % dq_capacity;
+		deque_shift(vm, dest, dq_capacity,
+					/* from */ dq_start,
+					/* to   */ index_val-1 + src_length-1,
+					src_length);
+	}
+	// else in right half, shift suffix elements right
+	else
+	{
+		deque_shift(vm, dest, dq_capacity,
+					/* from */ dq_start + dq_size + src_length - 1,
+					/* to   */ dq_start + index_val + src_length,
+					-src_length);
+	}
+	// insert elements in position
+	uint16 i;
+	for (i = 0; i < src_length; i++)
+		vm->variables[dest + 2 + ((dq_start + index_val + i) % dq_capacity)] = vm->variables[src + i];
+	vm->variables[dest] = dq_size = dq_size + src_length;
+}
+
+void AsebaNative_deqinsert(AsebaVMState *vm)
+{
+	// variable pos
+	uint16 dest = AsebaNativePopArg(vm);
+	uint16 src = AsebaNativePopArg(vm);
+	uint16 index = AsebaNativePopArg(vm);
+	uint16 index_val = vm->variables[index];
+	
+	// variable size
+	uint16 dest_length = AsebaNativePopArg(vm);
+	uint16 src_length = AsebaNativePopArg(vm);
+	
+	// Check for exception
+	if (vm->variables[index] < 0 || index_val > dest_length - 2 - 1
+		|| src_length > dest_length - 2 - vm->variables[dest])
```

Comment:
After thinking for a while, I understand that `vm->variables[dest]` is the index. I think that it would make the code more readable to put it explicitly in a temporary variable. But maybe a comment is sufficient.